subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CleanupOutputsStep.java

```diff
@@ -88,6 +92,8 @@ private void cleanupOverlappingOutputs(BeforeExecutionContext context, UnitOfWor
             );
             for (FileCollectionFingerprint fileCollectionFingerprint : previousOutputs.getOutputFileProperties().values()) {
                 try {
+                    // Previous outputs can be in a different place than the current outputs
+                    outputChangeListener.beforeOutputChange(fileCollectionFingerprint.getRootHashes().keySet());
```

Comment:
It seems we are using `FCF.getRootHashes().keySet()` in different places where we actually mean `FCF.getRootPaths()`. Maybe add a method like that?

Reply:
I added the method as part of https://github.com/gradle/gradle/pull/11819.